#### Дмитрий, привет! 

Меня зовут Максим Попов, и я буду проверять твой проект. Давай будем общаться на «ты». Но если тебе комфортнее на «вы», обязательно сообщи мне об этом. 
По ходу тетрадки ты найдешь комментарии к проекту, они будут в отдельных ячейках Markdown с заголовком «Комментарий ревьюера». Постарайся учесть эти комментарии для дальнейших проектов.

---

Я буду отмечать комментарии цветом и примечанием. В конце заголовка комментария будет указан номер итерации проверки. Пожалуйста, не удаляй и не правь мои комментарии

Критично - требует исправления, влияет на удачное выполнение проекта

Рекомендация - комментарий является рекомендацией или советом. Mожешь учесть их при выполнении будущих заданий. На твое усмотрение

Отлично - Так отмечены удачные решения и подходы, на которые стоит опираться в будущих проектах


---
<div class="alert alert-block alert-danger">
    

### Комментарий от ревьюера
Критично

Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.
</div>
    
---

<div class="alert alert-block alert-warning">
    

### Комментарий от ревьюера
Рекомендация

Такой комментарий является рекомендацией или советом. Mожешь учесть их при выполнении будущих заданий. На твое усмотрение. 
</div>

---

<div class="alert alert-block alert-success">
    

### Комментарий от ревьюера
Отлично
    
Так отмечены удачные решения и подходы, на которые стоит опираться в будущих проектах
</div>


Желательно реагировать на комментарии ('исправлено', 'не понятно как исправить ошибку' и т.п.) - так мы сможем более эффективно поработать над этим проектом.

---

Будет здорово, если ты будешь помечать свои действия следующим образом:



<div class="alert alert-block alert-info">

### Комментарий студента:

</div>

<div class="alert alert-block alert-success">
    

### Комментарий от ревьюера

Дмитрий, ты блестяще разбираешься с запросами, 5 из 5 с первого раза 👍🏻 .
    
Обращу внимание на следующие моменты:
    
- Оставляй комментарии на подзапросы, такая привычка поможет тебе, когда придется возвращаться к старому коду
    
- Попробуй заворачивать подзапросы в CTE. Так намного проще и читать, и писать код.

Если у тебя пока небольшой опыт работы с SQL, для дальнейшего изучения рекомендую следующие ресурсы:
    
1. https://datalearn.ru/kurs-po-sql 
    
Сразу переходи к оконкам и MS SQL Server, там много полезной инфы будет уже про создание таблиц, хранимые процедуры и т. п. 
    
2. https://stepik.org/course/3203/promo (это уже если хочется изучить поглубже)
    
Успеха в поиске работы!
            
</div>

### Введение

Цель исследования: проанализировать базу данных сервиса для чтения книг. Решить задачи и сделать выводы по каждому пункту

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], db_config['pwd'], db_config['host'], db_config['port'],
db_config['db']) 
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Просмотр таблиц

In [2]:
query_b = ''' SELECT *
            FROM books
            LIMIT 5
        '''

In [3]:
books = pd.io.sql.read_sql(query_b, con = engine)

In [4]:
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query_a = ''' SELECT *
            FROM authors
            LIMIT 5
        '''

In [6]:
authors = pd.io.sql.read_sql(query_a, con = engine)

In [7]:
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
query_p = ''' SELECT *
            FROM publishers
            LIMIT 5
        '''

In [9]:
publishers = pd.io.sql.read_sql(query_p, con = engine)

In [10]:
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
query_r = ''' SELECT *
            FROM ratings
            LIMIT 5
        '''

In [12]:
ratings = pd.io.sql.read_sql(query_r, con = engine)

In [13]:
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [14]:
query_rv = ''' SELECT *
            FROM reviews
            LIMIT 5
        '''

In [15]:
reviews = pd.io.sql.read_sql(query_rv, con = engine)

In [16]:
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-block alert-success">
    

### Комментарий от ревьюера
Отлично
    
Доступ к базе получили, на таблицы посмотрели, переходим к основной работе
    
</div>

### Задание 1. Посчитайте, сколько книг вышло после 1 января 2000 года

In [17]:
query1 = ''' SELECT COUNT(book_id)
            FROM books
            WHERE DATE(publication_date) > '2000-1-1'
        '''

In [18]:
books_cnt = pd.io.sql.read_sql(query1, con = engine)

In [19]:
books_cnt

count
0    819

После 1 января 2000 года выпущено 819 книг

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

    
👍🏻
</div>

### Задание 2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [41]:
query2 = ''' 

SELECT b.title AS book_name,
       b.book_id AS id,
       rv.n_rev AS reviews,
       r.avg_rat AS avg_rating
FROM books as b LEFT JOIN 
                (SELECT
                    book_id,
                    AVG(rating) AS avg_rat
                FROM
                    ratings
                GROUP BY 
                    book_id
                ) AS r ON r.book_id = b.book_id
        LEFT JOIN
                (SELECT 
                    book_id,
                    COUNT(review_id) AS n_rev
                FROM
                    reviews
                GROUP BY 
                    book_id
                ) AS rv ON rv.book_id = b.book_id
                
; ''' 

In [42]:
book_ratings_rewievs = pd.io.sql.read_sql(query2, con = engine)

In [43]:
book_ratings_rewievs

book_name   id  reviews  \
0            The Body in the Library (Miss Marple  #3)  652      2.0   
1                                            Galápagos  273      2.0   
2                             A Tree Grows in Brooklyn   51      5.0   
3    Undaunted Courage: The Pioneering First Missio...  951      2.0   
4                                          The Prophet  839      4.0   
..                                                 ...  ...      ...   
995                                Alice in Wonderland   64      4.0   
996          A Woman of Substance (Emma Harte Saga #1)   55      2.0   
997                                          Christine  148      3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)  790      2.0   
999                           The Plot Against America  828      2.0   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.285714  
..          ...  
995    4.230769  
996    5.000000  
997    3.428571  
998    3.500000  
999    3.000000  

[1000 rows x 4 columns]

Вывел данные по оценкам и обзорам для всех книг

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

👍🏻
    

Можно обойтись без подзапросов и приджойнить таблицы, если считать количество по уникальным обзорам COUNT(DISTINCT review_id)
    
И пропуски в количестве обзоров лучше заменить на 0, можно сделать с помощью COALESCE
    
</div>

### Задание 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [23]:
query3 = ''' 
SELECT bp.publisher_id,
       bp.publisher,
       bp.n_books
FROM
    (SELECT 
        b.publisher_id AS publisher_id,
        COUNT(b.book_id) AS n_books,
        p.publisher AS publisher 
    FROM 
        books as b
        LEFT JOIN publishers as p ON b.publisher_id=p.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        b.publisher_id,
        p.publisher
    ORDER BY 
        n_books DESC) AS bp
LIMIT 1
; ''' 

In [24]:
best_publisher = pd.io.sql.read_sql(query3, con = engine)

In [25]:
best_publisher

publisher_id      publisher  n_books
0           212  Penguin Books       42

Быльше всего книг выпустило издательство Penguin Books, 42 книги

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

👍🏻
    
</div>

### Задание 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [35]:
query4 = ''' 
SELECT a_avg.author,
       AVG(a_avg.avg_rating) AS avg_avg_rating
FROM
        (SELECT 
            a.author AS author,
            AVG(r.rating) AS avg_rating
        FROM books as b
            LEFT JOIN authors as a ON b.author_id = a.author_id
            LEFT JOIN ratings as r ON b.book_id = r.book_id
        GROUP BY 
             b.book_id,
             a.author_id
        HAVING
            COUNT(r.rating_id) >=50 
        ) AS a_avg 
GROUP BY a_avg.author
ORDER BY avg_avg_rating DESC
LIMIT 5
    
; ''' 

In [36]:
best_author = pd.io.sql.read_sql(query4, con = engine)

In [37]:
best_author

author  avg_avg_rating
0         J.K. Rowling/Mary GrandPré        4.283844
1  Markus Zusak/Cao Xuân Việt Khương        4.264151
2                     J.R.R. Tolkien        4.258446
3                  Louisa May Alcott        4.192308
4                       Rick Riordan        4.080645

Лучшим автором с самой высокой средней оценкой книг является J.K. Rowling/Mary GrandPré

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

👍🏻
    
Здесь частенько ошибаются, молодец, что считаешь именно книги с 50-ю и более оценками
    
</div>

### Задание 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [29]:
query5 = ''' 

SELECT
    AVG(mean_rewievs.n_reviews)
FROM
    (SELECT
        username,
        COUNT(review_id) AS n_reviews
    FROM
        reviews
    WHERE 
        username IN (
                SELECT 
                    username
                FROM 
                    ratings
                GROUP BY 
                    username
                HAVING 
                    COUNT(rating_id) > 50
                    )
    GROUP BY 
        username
    ) AS mean_rewievs
; ''' 

In [30]:
mean_rewievs = pd.io.sql.read_sql(query5, con = engine)

In [31]:
mean_rewievs

avg
0  24.333333

Среднее количество обзоров от пользователей составило 24.3

<div class="alert alert-block alert-success">
 

### Комментарий от ревьюера v_1

👍🏻
      
</div>

<div class="alert alert-block alert-warning">

### Комментарий от ревьюера v_1
Рекомендация

Работу следует подытожить общим выводом
</div>